### ISSUE TITLE
Better documentation for Sockeye + horovod

### ISSUE THREAD

**Issue by bricksdont** [awslabs/sockeye#942]
In my opinion, there is almost no documentation on how to use horovod for Sockeye training. It would be great if that could be added on Github or in the docs.

Examples for things I am not sure about:
- If horovod allows to specify specific GPU ids, or simply the number of devices
- If settings like the device for validation are relevant for horovod, or if it decides on its own where to run validation
- <strike>If horovod respects `CUDA_VISIBLE_DEVICES` in practice </strike>(it does)
- What primary and secondary workers are and which of them do validation

Thanks!
Mathias

**Comment by tdomhan** [comment]
Hi Mathias,
in general with Horovod it's best to let Horovod take care of multiple GPUs by creating one Sockeye process per GPU. For example you could use `--device-ids -1` per process and have horovod spawn say 8 processes:
```
horovodrun -np 8 python3 -m sockeye.train ...
```
You can still use negative device ids to let Sockeye automatically pick the GPU id.

For validation scores: all processes will run the evaluation (with the exception of the checkpoint decoder which runs only on the main process, see [here](https://github.com/awslabs/sockeye/blob/master/sockeye/train.py#L216)). As they will have the same parameters that are synced through Horovod they will get the same evaluation results.

I'll let @mjdenkowski also chime in as he's the Horovod expert.

Tobi

**Comment by mjdenkowski** [comment]
Hi Mathias,

Generally the only change required to run Horovod training is [specifying the per-device batch size](https://github.com/awslabs/sockeye/blob/795ea2746720adb8658b0f78eb475ba46b9c75ad/sockeye/arguments.py#L961-L967) instead of the total batch size.  Sockeye automatically handles the rest.

To answer your questions:

- When using Horovod, you don't need to specify any device IDs.  The N processes will use the N first available GPUs.  It is possible to specify a negative value to [run each process on multiple GPUs](https://github.com/awslabs/sockeye/blob/795ea2746720adb8658b0f78eb475ba46b9c75ad/sockeye/utils.py#L325-L328), but this is not recommended.  In practice, it is most efficient to use one GPU per process and have Horovod manage synchronization across GPUs.

- As Tobi pointed out, all processes run validation normally, though only the first process runs the checkpoint decoder.  The first process makes an authoritative check for improvement and broadcasts the result to the other processes so that decisions like reducing the learning rate or stopping training happen at the same time for all processes.

- The first process is referred to as the primary worker and other processes as secondary workers.  Each worker is an independent, fully fledged sockeye.train process that uses Horovod to sync data with other workers at defined points.  Because we don't need 8+ copies of everything, secondary workers do not keep full checkpoint histories and there are options for controlling their verbosity during training.

For a good balance between speed and cost, we recommend running Horovod training on a single host with 8 GPUs (see Tobi's `horovodrun` example).  This is the setup we used to train models for the [AMTA 2020 paper](https://arxiv.org/pdf/2008.04885.pdf).

Best,
Michael

**Comment by bricksdont** [comment]
Thanks Tobi and Michael!

By "validation" I meant the checkpoint decoder process - sorry that was not clear.
